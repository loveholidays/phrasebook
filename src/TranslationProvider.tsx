import React, { createContext, useContext, useMemo } from 'react';
import { DEFAULT_NAMESPACE } from './constants';

import { processTranslation } from './processTranslation';
import type {
  Namespaces, Locale, TFunction, TranslationData,
} from './types';

export interface TranslationContextValue {
  locale: Locale;
  namespaces: Namespaces;
  t: TFunction;
}

export const TranslationContext = createContext<TranslationContextValue>({
  locale: '',
  namespaces: {},
  t: () => '',
});

export interface ReplaceArgumentErrorParams {
  key: string;
  argumentName: string;
}

export type OnErrorCallback =
  & ((errorType: 'REPLACE_ARGUMENT_NOT_PASSED', params: ReplaceArgumentErrorParams) => void);

interface TranslationProviderProps {
  locale: Locale;
  namespaces?: Namespaces;
  translations?: TranslationData;
  onError?: OnErrorCallback;
}

// @TODO:
// - option for deep merge?
// - option for providing a custom `mergeNamespaces` function?
const mergeNamespaces = (
  parentNamespaces: Namespaces = {},
  namespaces: Namespaces = {},
  translations?: TranslationData,
): Namespaces => ({
  ...parentNamespaces,
  ...namespaces,
  ...(translations && {
    [DEFAULT_NAMESPACE]: translations,
  }),
});

export interface UseTranslationReturnValue {
  t: TFunction;
}

export const useTranslation = (namespace?: string) => ({
  t: namespace
    ? (key, args) => useContext(TranslationContext).t(key, {
      ns: namespace,
      ...args,
    })
    : useContext(TranslationContext).t,
}) as UseTranslationReturnValue;

/**
 * TranslationProvider
 * Provider used to create the localisation context.
 * @param {object} props
 * @param {string} props.locale String is used for locale specific formatting.
 * @param {object} props.namespaces Namespaced translations where the keys are the names of the namespaces and the values are the translations for the given namespace.
 * @param {object} props.translations Translation for default locale.
 * @param {function} props.onError Callback could be used to track the error during translation processing.
 * */
export const TranslationProvider: React.FC<React.PropsWithChildren<TranslationProviderProps>> = ({
  locale,
  namespaces,
  translations,
  onError,
  children,
}) => {
  const { namespaces: parentNamespaces } = useContext(TranslationContext);
  const mergedNamespaces = useMemo(
    () => mergeNamespaces(parentNamespaces, namespaces, translations),
    [ parentNamespaces, namespaces, translations ],
  );

  return (
    <TranslationContext.Provider
      value={{
        locale,
        namespaces: mergedNamespaces,
        t: (key, args) => processTranslation({
          locale,
          namespaces: mergedNamespaces,
          key,
          args,
          onError,
        }),
      }}
    >
      {children}
    </TranslationContext.Provider>
  );
};

