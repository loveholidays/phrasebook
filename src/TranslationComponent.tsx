import React, { Fragment, ReactElement } from 'react';

import { useTranslation } from './TranslationProvider';

type TranslationComponent = (ReactElement | ((text: string) => ReactElement));

const regExp = {
  singleComponent: /^<(\d+)>$/,
  childrenComponent: /^<(\d+)>([^<]*)<\/\d+>$/,
  extractComponents: /(<\d+>[^<]*<\/\d+>|<\d+>)/,
};

const resolveParts = (parts: string[], components: TranslationComponent[]) => parts.map((part) => {
  // <1>
  if (regExp.singleComponent.test(part)) {
    const [ , i ] = part.match(regExp.singleComponent)!;
    const index = Number(i) - 1;

    return components[index];
  }

  // <1>text</1>
  if (regExp.childrenComponent.test(part)) {
    const [ , i, text ] = part.match(regExp.childrenComponent)!;
    const index = Number(i) - 1;

    const component = components[index];

    if (typeof component === 'function') {
      return component(text);
    }

    return component;
  }

  return part;
});

interface TranslationProps {
  translationKey: string;
  namespace?: string;
  params?: Record<string, string | number | undefined>;
  components?: TranslationComponent[];
}

export const Translation: React.FC<TranslationProps> = ({
  translationKey,
  namespace,
  params = {},
  components = [],
}) => {
  const { t } = useTranslation(namespace);
  const translation = t(translationKey, params);
  const parts = translation.split(regExp.extractComponents);
  const resolvedParts = resolveParts(parts, components);

  return (
    <Fragment>
      {resolvedParts.map((part, i) => (
        <Fragment key={i}>
          {part}
        </Fragment>
      ))}
    </Fragment>
  );
};
