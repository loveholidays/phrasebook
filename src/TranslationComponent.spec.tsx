/* eslint-disable global-require */

import React from 'react';

import { render } from '@testing-library/react';
import { Translation } from './TranslationComponent';
import { TranslationProvider } from './TranslationProvider';
import namespaces from './testing/testNamespaces.json';

describe('Translation component', () => {
  const locale = 'en-GB';

  it('translates the default component', () => {
    const { asFragment } = render(
      <TranslationProvider
        locale={locale}
        namespaces={namespaces}
      >
        <Translation
          translationKey="translationComponent"
          params={{
            first: 'a',
            second: 'b',
          }}
          components={[
            <strong key={1}>1234</strong>,
            <strong key={2}>Y</strong>,
          ]}
        />
      </TranslationProvider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('translates with one param and two components', () => {
    const { asFragment } = render(
      <TranslationProvider
        locale={locale}
        namespaces={namespaces}
      >
        <Translation
          translationKey="oneParamTwoComponents"
          params={{
            first: 'a',
          }}
          components={[
            <strong key={1}>1234</strong>,
            <strong key={2}>Y</strong>,
          ]}
        />
      </TranslationProvider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('translates with two params and one component', () => {
    const { asFragment } = render(
      <TranslationProvider
        locale={locale}
        namespaces={namespaces}
      >
        <Translation
          translationKey="twoParamsOneComponent"
          params={{
            first: 'a',
            second: 'b',
          }}
          components={[
            <strong key={1}>1234</strong>,
          ]}
        />
      </TranslationProvider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('translates when the components have children', () => {
    const { asFragment } = render(
      <TranslationProvider
        locale={locale}
        namespaces={namespaces}
      >
        <Translation
          translationKey="componentWithChildren"
          components={[
            (text) => <a href="/#">{text}</a>,
            (text) => <strong>{text}</strong>,
          ]}
        />
      </TranslationProvider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
