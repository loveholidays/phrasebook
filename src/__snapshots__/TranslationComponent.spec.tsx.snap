// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Translation component translates the default component 1`] = `
<DocumentFragment>
  param first a param second b component one 
  <strong>
    1234
  </strong>
   component 2 
  <strong>
    Y
  </strong>
   something at the end
</DocumentFragment>
`;

exports[`Translation component translates using the provided namespace  1`] = `
<DocumentFragment>
  a _ b 
  <strong>
    @
  </strong>
   c
</DocumentFragment>
`;

exports[`Translation component translates when the components have children 1`] = `
<DocumentFragment>
  this makes it possible to 
  <a
    href="/#"
  >
    make some parts
  </a>
   of the 
  <strong>
    text
  </strong>
   appear as links
</DocumentFragment>
`;

exports[`Translation component translates with one param and two components 1`] = `
<DocumentFragment>
  param first a component one 
  <strong>
    1234
  </strong>
   component 2 
  <strong>
    Y
  </strong>
   something at the end
</DocumentFragment>
`;

exports[`Translation component translates with two params and one component 1`] = `
<DocumentFragment>
  param first a param second b component one 
  <strong>
    1234
  </strong>
   something at the end
</DocumentFragment>
`;
